facet-toml/src/error.rs

```diff
@@ -1,25 +1,175 @@
+        let label = Label::new((source_id, span.clone()))
+            .with_message(self.message())
+            .with_color(Color::Red);
+
+        report = report.with_label(label);
+
+        // Define the TOML source code
+        let source = Source::from(self.toml);
+
+        // Write to string
+        let mut writer = Vec::new();
+        if let Err(e) = report.finish().write((source_id, &source), &mut writer) {
+            return write!(f, "Error formatting with ariadne: {e}");
+        }
+
+        if let Ok(output) = String::from_utf8(writer) {
+            write!(f, "{}", output)
+        } else {
+            write!(f, "Error converting ariadne output to string")
+        }
     }
 }
 
-impl std::error::Error for AnyErr {}
+#[cfg(feature = "rich-diagnostics")]
+impl core::error::Error for TomlError<'_> {}
 
-impl From<String> for AnyErr {
-    fn from(s: String) -> Self {
-        Self(s)
+impl core::fmt::Debug for TomlError<'_> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        core::fmt::Display::fmt(self, f)
     }
 }
 
-impl From<&str> for AnyErr {
-    fn from(s: &str) -> Self {
-        Self(s.to_string())
+/// Type of error.
+#[derive(Debug, PartialEq)]
+pub enum TomlErrorKind {
+    /// Any error from facet.
+    GenericReflect(ReflectError),
+    /// Parsing TOML document error.
+    GenericTomlError(String),
+    /// Parsing a TOML type as a Rust type failed.
+    FailedTypeConversion {
+        /// TOML type that failed to convert.
+        toml_type_name: &'static str,
+        /// Rust that type didn't match the TOML type.
+        rust_type: &'static Shape,
+        /// Explanation why it failed.
+        reason: Option<String>,
+    },
+    /// Expected a certain TOML type, but got something else.
+    ExpectedType {
+        /// TOML type that was expected.
+        expected: &'static str,
+        /// TOML type that we got.
+        got: &'static str,
+    },
+    /// Found a TOML type that we don't know how to handle.
+    UnrecognizedType(&'static str),
+    /// Found a Rust scalar type that we don't know how to handle.
+    UnrecognizedScalar(&'static Shape),
+    /// Rust value is not a valid key.
+    InvalidKey(&'static Shape),
+    /// Expected a TOML field with the specified name, but couldn't find it.
+    ExpectedFieldWithName(&'static str),
+    /// Expected at least one field, got zero.
+    ExpectedAtLeastOneField,
+    /// Expected a single value, got multiple field.
+    ExpectedExactlyOneField,
+    /// Tried parsing a single value as a struct with multiple fields.
+    ParseSingleValueAsMultipleFieldStruct,
+}
+
+/// Wrap the `Shape` in a type so we can write it's type name directly with display.
+struct TypeNameWriter(&'static Shape);
+
+impl core::fmt::Display for TypeNameWriter {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.write_type_name(f, TypeNameOpts::default())
```

Comment:
@fasterthanlime I feel this was a bit cumbersome to find out, do you mind if I add a `type_name_to_string()` or similar named to `Shape`?

Reply:
> do you mind if I add a `type_name_to_string()` or similar named to `Shape`?

I do mind, `to_string()` allocates x)

But you can just do `{shape}` in error messages, since it implements Display.

Reply:
If I do `{shape}` I get the whole shape description, which is quite noisy for a single line. I only want to know the type name.

Reply:
You're always one step ahead of me, I was only looking at the `{shape:?}` implementation, thanks!